File: examples/go/p2pkh/P2PKH_test.go
package contract

import (
	"testing"

	runar "github.com/icellan/runar/packages/runar-go"
)

func TestP2PKH_Unlock(t *testing.T) {
	pk := runar.MockPubKey()
	c := &P2PKH{PubKeyHash: runar.Hash160(pk)}
	c.Unlock(runar.MockSig(), pk)
}

func TestP2PKH_Unlock_WrongKey(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected assertion failure for wrong public key")
		}
	}()
	pk := runar.MockPubKey()
	wrongPk := runar.PubKey("\x03" + string(make([]byte, 32)))
	c := &P2PKH{PubKeyHash: runar.Hash160(pk)}
	c.Unlock(runar.MockSig(), wrongPk)
}

func TestP2PKH_Compile(t *testing.T) {
	if err := runar.CompileCheck("P2PKH.runar.go"); err != nil {
		t.Fatalf("Rúnar compile check failed: %v", err)
	}
}

// Row 483: P2PKH is stateless — no mutable state tracked
func TestP2PKH_IsStateless(t *testing.T) {
	pk := runar.MockPubKey()
	c := &P2PKH{PubKeyHash: runar.Hash160(pk)}
	// Stateless contracts have no AddOutputs tracking
	// Calling Unlock does not accumulate state
	c.Unlock(runar.MockSig(), pk)
	// After the call, no side-effects on the struct itself
	// (the contract is stateless — properties are readonly)
	if len(c.PubKeyHash) == 0 {
		t.Error("expected PubKeyHash to remain non-empty after unlock")
	}
}

File: examples/rust/p2pkh/P2PKH_test.rs
#[path = "P2PKH.tsop.rs"]
mod contract;

use contract::*;
use tsop::prelude::*;

#[test]
fn test_unlock() {
    let pk = mock_pub_key();
    let c = P2PKH { pub_key_hash: hash160(&pk) };
    c.unlock(&mock_sig(), &pk);
}

#[test]
#[should_panic]
fn test_unlock_wrong_key() {
    let pk = mock_pub_key();
    let wrong_pk = vec![0x03; 33];
    let c = P2PKH { pub_key_hash: hash160(&pk) };
    c.unlock(&mock_sig(), &wrong_pk);
}

#[test]
fn test_compile() {
    tsop::compile_check(include_str!("P2PKH.tsop.rs"), "P2PKH.tsop.rs").unwrap();
}
